components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java

```diff
@@ -607,17 +610,45 @@ protected void processAuthenticationResponse(HttpServletRequest request,
-                userId = resolvedUserResult.getUser().getUserID();
-            } else {
-                context.setProperty(IS_INVALID_USERNAME, true);
-                throw new InvalidCredentialsException(ErrorMessages.USER_DOES_NOT_EXISTS.getCode(),
-                        ErrorMessages.USER_DOES_NOT_EXISTS.getMessage(), User.getUserFromUserName(username));
+            try {
+                List<String> userStorePreferenceOrder = getUserStorePreferenceOrder();
+                String userStoreQualifiedUsername = StringUtils.EMPTY;
+                if (userStorePreferenceOrder.size() == 1) {
+                    String allowedDomainName = userStorePreferenceOrder.get(0);
+                    userStoreQualifiedUsername = UserCoreUtil.addDomainToName(tenantAwareUsername, allowedDomainName);
+                }
+
+                if (userStorePreferenceOrder.isEmpty() || userStorePreferenceOrder.size() == 1) {
+                    ResolvedUserResult resolvedUserResult = null;
+                    if (userStorePreferenceOrder.isEmpty()) {
+                        resolvedUserResult = BasicAuthenticatorDataHolder.getInstance().getMultiAttributeLogin().
+                                resolveUser(tenantAwareUsername, requestTenantDomain);
+                    }
+                    if (userStorePreferenceOrder.size() == 1) {
+                        resolvedUserResult = BasicAuthenticatorDataHolder.getInstance().getMultiAttributeLogin().
+                                resolveUser(userStoreQualifiedUsername, requestTenantDomain);
+                    }
+                    if (resolvedUserResult != null && ResolvedUserResult.UserResolvedStatus.SUCCESS.
+                            equals(resolvedUserResult.getResolvedStatus())) {
+                        tenantAwareUsername = resolvedUserResult.getUser().getUsername();
+                        username = UserCoreUtil.addTenantDomainToEntry(tenantAwareUsername, requestTenantDomain);
+                        userId = resolvedUserResult.getUser().getUserID();
+                    } else {
+                        context.setProperty(IS_INVALID_USERNAME, true);
+                        throw new InvalidCredentialsException(ErrorMessages.USER_DOES_NOT_EXISTS.getCode(),
+                                ErrorMessages.USER_DOES_NOT_EXISTS.getMessage(), User.getUserFromUserName(username));
+                    }
+                } else {
+                    throw new InvalidCredentialsException(
+                            ErrorMessages.MULTIPLE_USER_STORE_BINDING_FOR_SP_NOT_ALLOWED.getCode(),
+                            ErrorMessages.MULTIPLE_USER_STORE_BINDING_FOR_SP_NOT_ALLOWED.getMessage(),
+                            User.getUserFromUserName(username));
+                }
+            } catch (UserStoreException e) {
+                log.error(ErrorMessages.USER_STORE_EXCEPTION_WHILE_TRYING_TO_AUTHENTICATE.getMessage());
+                throw new AuthenticationFailedException(
```

Comment:
```suggestion
                List<String> userStorePreferenceOrder = getUserStorePreferenceOrder();
                // Add a comment which situation userStorePreferenceOrder will have the size 1
                if (userStorePreferenceOrder.size() == 1) {
                    String allowedDomainName = userStorePreferenceOrder.get(0);
                    tenantAwareUsername = UserCoreUtil.addDomainToName(tenantAwareUsername, allowedDomainName);
                }

              
                        resolvedUserResult = BasicAuthenticatorDataHolder.getInstance().getMultiAttributeLogin().
                                resolveUser(tenantAwareUsername, requestTenantDomain);
                   
                   
                    if (resolvedUserResult != null && ResolvedUserResult.UserResolvedStatus.SUCCESS.
                            equals(resolvedUserResult.getResolvedStatus())) {
                        tenantAwareUsername = resolvedUserResult.getUser().getUsername();
                        username = UserCoreUtil.addTenantDomainToEntry(tenantAwareUsername, requestTenantDomain);
                        userId = resolvedUserResult.getUser().getUserID();
                    } else {
                        context.setProperty(IS_INVALID_USERNAME, true);
                        throw new InvalidCredentialsException(ErrorMessages.USER_DOES_NOT_EXISTS.getCode(),
                                ErrorMessages.USER_DOES_NOT_EXISTS.getMessage(), User.getUserFromUserName(username));
                    }
               
```

Reply:
Can't we go ahead with the above suggested approach. If the user userstore preference order is one then we will check within the user store other wise it will behave as below(how the flow works without the feature)

Reply:
Resolved with 58295f3237522fdd2dc1e090a3c7a6b0778fced4